google/cloud/testing_util/validate_metadata.cc

```diff
@@ -87,15 +87,12 @@ RoutingHeaders ExtractMDFromHeader(std::string header) {
  * `GTEST_USES_POSIX_RE`.
  */
 MATCHER_P(MatchesGlob, glob, "matches the glob: \"" + glob + "\"") {
-  // Translate the glob into a regex pattern.
-  auto matcher1 = absl::StrReplaceAll(glob, {{"*", "[^/]+"}});
-  // Create a second matcher that replaces all "/" with "%2F" to match the
-  // character replacement in `internal::UrlEncode`.
-  auto decoded_glob = absl::StrReplaceAll(glob, {{"/", "%2F"}});
-  auto matcher2 = absl::StrReplaceAll(decoded_glob, {{"*", "[^/]+"}});
-  auto regex_string = absl::StrCat(matcher1, "|", matcher2);
-  std::regex regex(regex_string);
-  return std::regex_match(arg, regex);
+  // Translate the `glob` into a regex pattern. Decode the glob.
+  auto matcher =
+      absl::StrReplaceAll(internal::UrlDecode(glob), {{"*", "[^/]+"}});
```

Comment:
I would still assert that `MatchesGlob` has no reason to accept url-encoded globs (and indeed that we would not want to support that), so we should just do the glob->regex transformation and leave it at that.  That is, drop the `UrlDecode()` here.

Reply:
Tests like ` bazel test google/cloud/bigtable:table_apply_test --test-output=all` fail when I remove it. Looking into it...
```
[  FAILED  ] TableApplyTest.Retry (482 ms)
[ RUN      ] TableApplyTest.RetryIdempotent
google/cloud/testing_util/validate_metadata.cc:345: Failure
Value of: actual
Expected: has 1 element and that element has a first field that is equal to "table_name", and has a second field that matches the glob: "projects%2Ffoo-project%2Finstances%2Fbar-instance%2Ftables%2Fbaz-table"
  Actual: { ("table_name", "projects%2Ffoo-project%2Finstances%2Fbar-instance%2Ftables%2Fbaz-table") }, where the following matchers don't match any elements:
matcher #0: has a first field that is equal to "table_name", and has a second field that matches the glob: "projects%2Ffoo-project%2Finstances%2Fbar-instance%2Ftables%2Fbaz-table"
and where the following elements don't match any matchers:
element #0: ("table_name", "projects%2Ffoo-project%2Finstances%2Fbar-instance%2Ftables%2Fbaz-table")
```

Reply:
> has a second field that matches the glob: "projects%2Ffoo-project%2Finstances%2Fbar-instance%2Ftables%2Fbaz-table"

So someone is url-encoding a glob?  Why would they choose to do that?

Reply:
It was probably me at some point.... I just wanted to make the tests pass. 

I think that this is a more involved change to find all the places that encode the globs and get rid of them. I propose tackling it in a second PR.

Reply:
They were probably compensating for the fact that `MatchesGlob` was not url-decoding the `arg`.  That is, it looks like `FromRoutingRule()` should not be url-encoding `match[1].str()`.

Reply:
Oh wow, you found that issue much faster than i would have. Thanks Brad!

Reply:
You could ask Darren as it looks like it originated in #12447.

There is also a `UrlEncode(field)` a few lines up that appears to fall into the same bucket.

Reply:
Just decoding the values is not enough. Things like `"projects/project"` start passing, when we want them to fail. (They should be `"projects%2Fproject"`).

We need something that verifies that the values were URL encoded to begin with. #12855 is good enough for me.